async/await support and other changes for meteor@3

dr-dimitru opened the pull request.

dr-dimitru commented:
@jankapunkt created this PR to track changes inherited from #884

jankapunkt commented:
@dr-dimitru so you think this is ready for an alpha release? We have two medium sized apps (100k loc) with lots of files involving which we can use to test

dr-dimitru commented:
@jankapunkt I'm working on couple of more changes, planning to pre-release today. Testing from your end would be very helpful

Friction-less-development commented:
@dr-dimitru Where is this pre-release? The latest I see is `3.0.0-beta.6` which does not seem to be up to date.

dr-dimitru commented:
@Friction-less-development @jankapunkt `v3.0.0-beta.7` is out, please give it a try

Friction-less-development commented:
Thank you, I was also happy to see it kept legacy count() on the front-end, alleviating some churn on that front.

I so far have seen one issue.

I am running the new function, this.removeAsync in the onAfterUpload hook.
(Basically doing some server-side checks, if they fail, I remove the file.)
But it doesn't seem to be working right. Through debugging, I can see when it reaches [this code](https://github.com/veliovgroup/Meteor-Files/blob/181fccac235655c5a82c43a41211762a3ed95a66/server.js#L1658-L1659), the files cursor is empty (And resolves to an empty array.)
So to fix, should either keep a copy of the collection for use when unlinking, or unlink earlier.

Because of this issue, while it is true that the collection item is being removed, the associated file is not being deleted from the server when removeAsync is called.

dr-dimitru commented:
@Friction-less-development thanks a lot for reporting this one, `ostrio:files@3.0.0-rc.2` is out with fix for `removeAsync` included, please give it a try

dr-dimitru commented:
@jankapunkt @bratelefant @harryadel @StorytellerCZ here's draft changelog for `ostrio:files@3.0.0-rc.3`, going to refactor and update docs now, please report issues if any.

### Summary

- ✨ Refactor: Hook options: `protected`, `onBeforeRemove`, `onAfterRemove`, `onInitiateUpload`, `onAfterUpload`, `namingFunction` are now *async*
- 🤝 Refactor: Compatibility with `meteor@3` and other modern packages
- ☄️ Refactor: Match `FilesCollection` APIs with new `*Async` methods of `Mongo.Collection`; Deprecate callback APIs on the Server
- 👨‍💻 Refactor: Utilize node's async APIs where suitable
- 👨‍💻 Refactor: Improve pause/resume logic on connection interruption/reconnect
- 📔 Docs: Updated and refactored docs with better examples
- 📔 Docs: Refactored JSDoc matching definitions in TypeScript
- 🤓 Dev: Improved TypeScript support
- 👷‍♂️ Dev: Improved debugging logs
- 👨‍🔬 Tests: Improved test-suite
- 👷‍♂️ Git: CI GitHub Action Workflows for lint and build tests

### Major changes

__FilesCollection__:
- ⚠️ `FilesCollection#remove()` — deprecated on server, use `FilesCollection#removeAsync` instead
- ⚠️ `FilesCollection#findOne()` — deprecated on server, use `FilesCollection#findOneAsync` instead
- ⚠️ `FilesCollection#unlink()` — deprecated on server, use `FilesCollection#unlinkAsync` instead
- ⚠️ `FilesCollection#write()` — deprecated on server, use `FilesCollection#writeAsync` instead
- ⚠️ `FilesCollection#load()` — deprecated on server, use `FilesCollection#loadAsync` instead

__FileCursor__:
- ⚠️ `FileCursor#remove()` — deprecated on server, use `FileCursor#removeAsync` instead

__FilesCursor__:
- ⚠️ `FilesCursor#remove()` — deprecated on server, use `FilesCursor#removeAsync` instead
- ⚠️ `FilesCursor#hasNext()` - deprecated, use `FilesCursor#hasNextAsync` instead
- ⚠️ `FilesCursor#count()` - deprecated, use `FilesCursor#countDocuments` instead
- ⚠️ `FilesCursor#countAsync()` - deprecated, use `FilesCursor#countDocuments` instead

__FileUpload__:
- ⚠️ `FileUpload#start()` is now *async*!

__Callbacks and hooks__:
- ⚠️ Anywhere: `this.user()` is deprecated, use `this.userAsync()` instead
- ⚠️ Client: `FileUpload` now always triggers `end` even in the case of successful and failed uploads; *Before: `end` event wasn't called under certain conditions*
- ⚠️ Client: All errors appeared during upload in all hooks and events of `FileUpload` are now instance of `Meteor.Error`; *Before: Errors had mixed type or were simply text*
- ⚠️ Client: Errors are the same now (type, code, text, reason, details) within DDP and HTTP protocols; *Before: DDP and HTTP protocols had different errors*
- ⚠️ Client: The next *private* events were removed from `UploadInstance` Class: `upload`, `sendEOF`, `prepare`, `sendChunk`, `proceedChunk`

### New methods

__FilesCollection__:
- ✨ Client: `FilesCollection#insertAsync()`
- ✨ Anywhere: `FilesCollection#updateAsync()`
- ✨ Anywhere: `FilesCollection#removeAsync()`
- ✨ Anywhere: `FilesCollection#findOneAsync()`
- ✨ Anywhere: `FilesCollection#countDocuments()`
- ✨ Anywhere: `FilesCollection#estimatedDocumentCount()`
- ✨ Server: `FilesCollection#unlinkAsync()`
- ✨ Server: `FilesCollection#writeAsync()`
- ✨ Server: `FilesCollection#loadAsync()`

__FileCursor__:
- ✨ Anywhere: `FileCursor#removeAsync()`
- ✨ Anywhere: `FileCursor#fetchAsync()`

__FilesCursor__:
- ✨ Anywhere: `FilesCursor#getAsync()`
- ✨ Anywhere: `FilesCursor#hasNextAsync()`
- ✨ Anywhere: `FilesCursor#nextAsync()`
- ✨ Anywhere: `FilesCursor#hasPreviousAsync()`
- ✨ Anywhere: `FilesCursor#previousAsync()`
- ✨ Anywhere: `FilesCursor#removeAsync()`
- ✨ Anywhere: `FilesCursor#fetchAsync()`
- ✨ Anywhere: `FilesCursor#firstAsync()`
- ✨ Anywhere: `FilesCursor#lastAsync()`
- ✨ Anywhere: `FilesCursor#countDocuments()`
- ✨ Anywhere: `FilesCursor#forEachAsync()`
- ✨ Anywhere: `FilesCursor#eachAsync()`
- ✨ Anywhere: `FilesCursor#mapAsync()`
- ✨ Anywhere: `FilesCursor#currentAsync()`
- ✨ Anywhere: `FilesCursor#observeAsync()`
- ✨ Anywhere: `FilesCursor#observeChangesAsync()`

### New features

- ✨ Client: `FileUpload#remainingTime` *ReactiveVar* — returns remaining upload time in `hh:mm:ss` format;

### Other changes

- 🔧 Security: Fixed #894 — now `x_mtok` cookie is set with `secure` and `httpOnly` flags
- 🔧 Refactor: Fixed `FileCursor#with` implementation
- ✨ Refactor: `FileUpload#abort` is now `async`
- ✨ Server: `FilesCollection#addFile` is now *async*
- ✨ Server: `FilesCollection#download` is now *async*
- ✨ Server: `WriteStream` Class is now available for import
- ⚙️ Client: `disableUpload` option processed on the client and returns error in the `end`/`error` events, and `onError` hooks. *Before — throws an error upon calling `.insert()` method*

### Dependencies

__release__:

- `eventemitter3@5.0.1`, *was `4.0.7`*
- *removed:* `abort-controller`, now using native `AbortController`
- *removed:* `fs-extra`, now using native `fs`

__dev__:

- *added:* `chai@4.5.0`
- *added:* `sinon@7.5.0`

harryadel commented:
I've just used the latest RC in our project. The only thing we had to change really was `unlink` to `unlinkAsync` which was about 3 calls. So far so good, I'll report back if I find any issues in the upcoming days. Thank you all for pushing this through.

dr-dimitru commented:
@harryadel awesome, thank you for feedback. Do you use `protected` other `this.user()` in any callback's contexts? Have some reports that this one is broken

harryadel commented:
>@harryadel awesome, thank you for feedback. Do you use protected other this.user() in any callback's contexts? Have some reports that this one is broken

No, I don't think we really do.

jankapunkt commented:
I am running a few tests right now with rc 3 and give you a summary tomorrow

jankapunkt commented:
Why is `onBeforeUpload` not async as well? This prevents me from doing db calls, such as checking for Roles with the involved user or other npm tools that are async by default.

harryadel commented:
@dr-dimitru I'm so glad you're back in the mix. Hopefully we can have a full release soon and get rid of the betas :clap:

dr-dimitru reviewed (commented):
lgtm
